ops/testing.py

```diff
@@ -1759,6 +1789,104 @@ class NonAbsolutePathError(Exception):
     """
 
 
+class _MockStorage:
+    def __init__(self, location, src):
+        self._src = src
+        os.makedirs(src, exist_ok=True)
+        self._location = location
+
+    def contains(self, path):
+        return str(path).startswith(self._location)
+
+    def _srcpath(self, path):
+        suffix = path[len(self._location):]
+        if suffix[0] == '/':
+            suffix = suffix[1:]
+        srcpath = os.path.join(self._src, suffix)
+        return srcpath
+
+    def create_dir(self, path: str, make_parents: bool = False, **kwargs) -> '_Directory':
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        dirname = os.path.dirname(srcpath)
+        if not os.path.exists(dirname):
+            if not make_parents:
+                raise RuntimeError('no such directory {}'.format(dirname))
+            os.makedirs(dirname)
+        os.makedirs(srcpath)
+        return _Directory(self, path, **kwargs)
+
+    def create_file(
+            self,
+            path: str,
+            data: typing.Union[bytes, str, typing.BinaryIO, typing.TextIO],
+            encoding: typing.Optional[str] = 'utf-8',
+            make_dirs: bool = False,
+            **kwargs
+    ) -> '_File':
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        dirname = os.path.dirname(srcpath)
+        if not os.path.exists(dirname):
+            if not make_dirs:
+                raise RuntimeError('no such directory {}'.format(dirname))
+            os.makedirs(dirname)
+
+        if not isinstance(data, bytes):
+            data = data.encode(encoding=encoding)
+
+        with open(srcpath, 'wb') as f:
+            f.write(data)
+
+        return _File(pathlib.PurePosixPath(path), data, encoding=encoding, **kwargs)
+
+    def list_dir(self, path) -> typing.List['_File']:
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        results = []
+        for fname in os.listdir(srcpath):
+            fpath = os.path.join(srcpath, fname)
+            mountpath = os.path.join(path, fname)
+            if os.is_dir(fpath):
+                results.append(_Directory(mountpath))
```

Comment:
_Directory actually defines its parameter as PurePosixPath not a plain str.
in current python
```
>>> import pathlib
>>> etc = pathlib.PurePosixPath('/etc')
>>> import os
>>> os.path.join(etc, 'foo')
'/etc/foo'
```
So os.path will accept pathlib, but they do always return a str.

Reply:
I like `etc / 'foo'` as well

Reply:
Done.  All converted to pathlib.